### ISSUE TITLE
Query access rule not hit if firewall's anonymous set to true

### ISSUE THREAD

**Issue by danaki** [overblog/GraphQLBundle#364]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | no
| Version/Branch   | v0.11.4

I'm trying to use GraphQL with JWT (lexik/jwt-authentication-bundle       v2.5.4), here are my configs:

security.yaml:
```yaml
security:
    firewalls:
        dev:
            pattern: ^/(_(profiler|wdt)|css|images|js)/
            security: false
        api:
            anonymous: true
            stateless: true
            guard:
                authenticators:
                    - lexik_jwt_authentication.jwt_token_authenticator

    access_control:
        - { path: ^/api, roles: [IS_AUTHENTICATED_FULLY, IS_AUTHENTICATED_ANONYMOUSLY] }
```

Query.types.yaml:
```yaml
Query:
    type: object
    config:
        description: "Query API"
        fields:
            myTransactions:
                type: "[TransactionObject]"
                args:
                    status:
                        type: "TransactionStatus"
                resolve: '@=resolver("App\\GraphQL\\Resolvers\\TransactionsResolver", [args])'
                access: "@=!isAnonymous()"
```

Notice the last "access: "@=!isAnonymous()"" line. I tried all of the directives like isAuthenticated(), isFullyAuthenticated() etc.

TransactionsResolver.php:
```php
<?php

declare(strict_types=1);

namespace App\GraphQL\Resolvers;

use App\Repository\TransactionRepository;
use Overblog\GraphQLBundle\Definition\Argument;
use Overblog\GraphQLBundle\Definition\Resolver\ResolverInterface;
use Symfony\Component\Security\Core\Security;

class TransactionsResolver implements ResolverInterface
{
    /**
     * @var TransactionRepository
     */
    private $transactionRepository;

    /**
     * @var Security
     */
    private $security;

    public function __construct(
        TransactionRepository $transactionRepository,
        Security $security
    ) {
        $this->transactionRepository = $transactionRepository;
        $this->security = $security;
    }

    public function __invoke(Argument $args)
    {
        $status = $args['status'] ?? null;

        return $this
            ->transactionRepository
            ->findByUserAndStatus($this->security->getUser(), $status);
    }
}
```

Now, if I'm not logged in and have no "Authorization: Bearer ..." header, this query:
```graphql
{
	transactions {
		status
	}
}
```

leads to this error:
```json
{
	"errors": [
		{
			"debugMessage": "Argument 1 passed to App\\Repository\\TransactionRepository::findByUserAndStatus() must be an instance of App\\Entity\\User, null given, called in \/Users\/yuri\/work\/backend\/src\/GraphQL\/Resolvers\/TransactionsResolver.php on line 38",
			"message": "Internal server Error",
			"category": "internal",
			"locations": [
				{
					"line": 2,
					"column": 2
				}
			],
			"path": [
				"transactions"
			],
```

The query resolver actually should not be called at all, but it's called while security user is null.

**Comment by danaki** [comment]
At the same time Mutation.types.yaml:
```
Mutation:
    type: object
    config:
        description: "Mutation API"
        fields:
            createTransaction:
                type: "String!"
                description: "Create a transaction with the current user"
                access: "@=isAuthenticated()"
                args:
                    input:
                        type: CreateTransactionInput!
                resolve: '@=mutation("App\\GraphQL\\Mutators\\CreateTransactionMutator", [args])'
```

called by anonymous user produces error:
```
{
	"errors": [
		{
			"message": "Access denied to this field.",
			"category": "user",
			"locations": [
				{
					"line": 8,
					"column": 2
				}
			],
			"path": [
				"createTransaction"
			],
```

which is expected behavior.

**Comment by Kocal** [comment]
https://github.com/overblog/GraphQLBundle/blob/master/Resources/doc/security/fields-access-control.md

> in query mode: execute resolver -> execute access -> manage result in function of access

Anyway, I'm not able to explain this behavior... maybe for executing voters on the resolved resource? 🤔 🤷‍♂️   
ping @mcg-web 

Would it be possible to add a field `accessMode` which will takes two values `beforeResolve` and `afterResolve` (default)?

**Comment by danaki** [comment]
Little update: actually access rule is hit, the title is now misleading. All returned objects from resolver now look like null to the caller if client is anonymous. However what I expected is that resolver should not be called at all. Can't find in docs how to achieve this behaviour, except to make some auth checks inside the resolver but this makes access rules described in several places, which is bad.

**Comment by mcg-web** [comment]
@danaki this is a known issue in query mode resolver is always execute since you can use `object` expression variable (the result of your resolver) to add criteria to access.  A solution right now is to directly inject the user using expression language and return null if user is not define.

 ```yaml
Query:
    type: object
    config:
        description: "Query API"
        fields:
            myTransactions:
                type: "[TransactionObject]"
                args:
                    status:
                        type: "TransactionStatus"
                resolve: '@=resolver("App\\GraphQL\\Resolvers\\TransactionsResolver", [getUser(), args])'
                access: "@=!isAnonymous()"
```
```php
<?php

declare(strict_types=1);

namespace App\GraphQL\Resolvers;

use App\Entity\User;
use App\Repository\TransactionRepository;
use Overblog\GraphQLBundle\Definition\Argument;
use Overblog\GraphQLBundle\Definition\Resolver\ResolverInterface;

class TransactionsResolver implements ResolverInterface
{
    /**
     * @var TransactionRepository
     */
    private $transactionRepository;

    public function __construct(
        TransactionRepository $transactionRepository
    ) {
        $this->transactionRepository = $transactionRepository;
    }

    public function __invoke( ?User $user, Argument $args)
    {
        if (null === $user) {
            return null;
        }

        return $this
            ->transactionRepository
            ->findByUserAndStatus($this->security->getUser(), $args['status']);
    }
}
```

**Comment by akomm** [comment]
@danaki try `composer update overblog/graphql-bundle`

If your repository returns the transaction as array, you might have been affected by this: #387 

It has been fixed up from version 0.9

**Comment by mcg-web** [comment]
Fixed, thanks again for feedback!
